File: project_euler/Problem052/Problem052.py
from typing import List

class Problem052:
    def solution(self) -> None:
        target, digits, index = 6, [[0 for _ in range(10)] for _ in range(2)], 10
        while True:
            self.__get_digits(index, digits[0])
            is_same = True
            for i in range(target, 1, -1):
                self.__get_digits(index * i, digits[1])
                if not self.__check_digits(digits):
                    is_same = False
                    break
            if is_same:
                print(index)
                return
            index += 1

    def __get_digits(self, value: int, digits: List[int]) -> None:
        digits[:] = [0 for _ in range(10)]
        while value:
            digits[value % 10] += 1
            value //= 10

    def __check_digits(self, digits: List[List[int]]) -> bool:
        for i in range(10):
            if digits[0][i] != digits[1][i]:
                return False
        return True

File: project_euler/Problem052/Problem052.go
package project_euler

import "fmt"

func Problem052() {
    target := 6
    digits := make([][]int, 2)
    for i := 0; i < 2; i++ {
        digits[i] = make([]int, 10)
    }
    index := 10
    for true {
        getDigits(index, digits[0])
        isSame := true
        for i := target; i > 1; i-- {
            getDigits(index * i, digits[1])
            if !checkDigits(digits) {
                isSame = false
                break
            }
        }
        if isSame {
            fmt.Println(index)
            return
        }
        index++
    }
}

func getDigits(value int, digits []int) {
    for i := 0; i < 10; i++ {
        digits[i] = 0
    }

    for 0 != value {
        digits[value % 10]++
        value /= 10
    }
}

func checkDigits(digits [][]int) bool {
    for i := 0; i < 10; i++ {
        if digits[0][i] != digits[1][i] {
            return false
        }
    }
    return true
}
